readme.txt
WiFiController v0.4.4
Copyright (c)2007 David Rudie (d.rudie@gmail.com)
http://imk.cx/psp/wificontroller/


READ ME.  READ ALL OF ME!


LICENSE
 This application uses the BSD license and a copy of the included Copyright
 LICENSE file can be found in the root of the zip file.


ABOUT THIS PROJECT
 Mikko Monone wrote the original controller program for the PSP and the server
 for the PC. After using his program on the PSP it seemed it was missing some
 things. The start button exited the program and select didn't work. It had a
 display for the buttons you would press, and while it was functional, I like
 a little more detail to be displayed. I then wrote my own controller from
 scratch and only borrowed a small portion of his code (the packet creation).
 Here is a link to the thread that started this all in a way for me.


AFFILIATION NOTE
 I don't know Mikko Monone or the authors of PPJoy. There is no affiliation
 with them at all. I thank them for their programs.


WHAT THIS IS AND WHAT THIS DOES
 This basically turns your PSP into a controller for your PC (running
 Windows). How it works is that PPJoy provides a way to create a virtual
 joystick. We can then communicate with this virtual joystick as though it
 were a real joystick. This is accomplished by running a server in Microsoft
 Windows. The server just runs in the background waiting for a connection.
 When you start the WiFiController program on your Sony PSP it will establish
 a network connection over wireless using infrastructure mode. Once it
 establishes a connection and obtains an IP address it will connect to the
 server running on your PC. Once it has connected to your PC it will act just
 like a regular joystick.


NOTES
 Although I have not comfirmed this myself, I have heard that this works on
 v2.60 PSP's as well. I originally wrote this program for v1.00 and v1.50 only.
 Recently, I have added official support for v3.xx OE.  When using v3.xx, make
 sure you run it with the v3.xx kernel.  If you use CWCheat and the v3.xx
 version of WiFiController, hold down the right trigger when launching the
 program.  If you do not, WiFiController will crash.


INSTALLATION
 I'm going to attempt to explain how to get this working. These steps are all
 necessary and you should pay close attention to them.

 1) Open the PSP directory and edit the proper wifi.cfg for your PSP version
    and set the IP of the computer you are going to use this on. If you don't
    know your IP address. Go to your "Start" menu. Click on "Run..." When the
    box opens up type "cmd.exe" without the quotes. When the command prompt
    opens type "ipconfig" without the quotes. It will report your IP address.

 2) Copy the correct director(y|ies) to your PSP depending on whether you have
    a v1.00, v1.50 or v3.xx OE PSP.

 3) Download and install PPJoy.
    http://www.geocities.com/deonvdw/Docs/PPJoyMain.htm

 4) After it is installed open up your "Control Panel" and click on "Parallel
    Port Joysticks".

 5) When the window comes up click "Add..." at the bottom. On the next screen
    change "Parallel port" to "Virtual joysticks" and then click "Add."

 6) It will want to install some drivers so either have it scan automatically
    or point it to the directory you installed PPJoy to.

 7) After it is done installing, click the "PPJoy Virtual joystick 1" in the
    menu and click "Mapping..."

 8) Choose "Set a custom mapping for this controller" and click "Next."

 9) Set these options for best results.
    Axes = 2 (X Axis and Y Axis)
    Buttons = 9
    POV hats = 1

 10) X Axis = Analog 0
     Y Axis = Analog 1

 11) These buttons are up to you to set. You should leave "Button 1" set to
     "nothing" otherwise you may have problems using the Triangle button on
     your PSP later.
     Button 1 = nothing
     Button 2 = Digital 0
     Button 3 = Digital 1
     Button 4 = Digital 2
     Button 5 = Digital 3
     Button 6 = Digital 4
     Button 7 = Digital 5
     Button 8 = Digital 11
     Button 9 = Digital 10

 12) Directional button POV
     North = Digital 6
     East = Digital 7
     West = Digital 9
     South = Digital 8

 13) Repeat steps 5 to 12 to add up to 4 joysticks (this is the maximum amount
     that WiFiController will support at the moment).

 14) Inside the PC directory, you will find a WiFiServer.exe.  If you are
     just using one PSP to play games, you can run this file directly.  If,
     however, you are or will be using more than one PSP, you will need to
     run one of the batch files in the directory.  Simply double-click the
     batch file for the appropriate controller.  You must have a server
     running for each controller you wish to use.  The server(s) must also be
     running at all times while using the PSP as a joystick.  When you are
     done using it, you can safely exit the program by pressing CTRL+C or
     clicking the close button for the window.

     Example:
       Two players will be playing a game.  You will need to start the batch
       files for controller 1 and 2.  Player 1 will choose Controller 1 from
       WiFiController and player 2 will choose Controller 2.

     Note:
       If the window opens and closes immediately, something went wrong and
       you'll need to either contact me or figure it out yourself.  However,
       make sure you RETRY all the steps before contacting me.  Also try
       reading over the WiFiController website and the PPJoy website for help.

 15) Now start the program on your PSP, choose your connection and your
     controller.  It should connect to your computer and basically be working.
     You'll find out in the next step.

 16) Go back to your Control Panel and go to Game Controllers. Click on "PPJoy
     Virtual joystick 1" and choose Properties. On the screen that comes up
     you should be able to press buttons on the PSP and have them show as
     working on the screen. I recommend calibrating the Analog Stick too.

 17) Enjoy!


CHANGES
v0.4.4)
  * Fixed so it will work under 3.40 kernel and higher.
  * Added a little bit more detail to the connection screen.

v0.4.3)
  * Renamed WiFiController.exe to WiFiServer.exe so it won't be confused for
    the actual PSP application itself.
  * Added the ability to act as up to 4 separate controllers.

v0.4.2)
  * Rewrote some functions to clean up code more.
  * Shortened all message displays by 1 second.

v0.4.1)
  * Fixed configuration selection menu.
  * Changed invert combo to L+R+SELECT.
  * Renamed SocketTest.exe to WiFiController.exe

v0.4)
  * Cleaned up code some more and removed old header files not used anymore.
  * Added ability to select which WLAN configuration to use.
  * L+R+START now inverts analog Y in case some games don't support it.
  * Probably new bugs introduced.  :)

v0.3)
  * Cleaned up and added more detailed instructions.
  * Fixed up on-screen display some.
  * Added some more detailed battery information.
  * Cleaned up the connection code some.
  * Moved some text strings to a new file.
  * Cleaned up code overall.


main.c
/*
 * Copyright (c) 2007 David Rudie
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <pspkernel.h>
#include <pspdebug.h>
#include <pspsdk.h>
#include <stdio.h>
#include <string.h>
#include <stdarg.h>
#include <unistd.h>
#include <pspwlan.h>
#include <pspnet.h>
#include <pspnet_inet.h>
#include <pspnet_apctl.h>
#include <psppower.h>
#include <arpa/inet.h>
#include <pspdisplay.h>
#include <pspctrl.h>
#include <psputility.h>
#include <psputility_netparam.h>

#include "defines.h"

# ifdef PSPFW3
PSP_MODULE_INFO(MODULE_NAME, 0x0, 0, 4);
PSP_HEAP_SIZE_KB(1024*16);
# else
PSP_MODULE_INFO(MODULE_NAME, 0x1000, 0, 4);
# endif
//PSP_MAIN_THREAD_ATTR(PSP_THREAD_ATTR_USER);
PSP_MAIN_THREAD_ATTR(0);
PSP_MAIN_THREAD_STACK_SIZE_KB(32);

int done = 0;
char confname[128];
extern int connect_to_apctl(int config);

// --------------------------------------------------------------------------


int ExitCallback(int arg1, int arg2, void *common) {
  done = 1;
  sceKernelExitGame();
  return 0;
}

// --------------------------------------------------------------------------

int CallbackThread(SceSize args, void *argp) {
  int cbid;

  cbid = sceKernelCreateCallback("Exit Callback", ExitCallback, NULL);
  sceKernelRegisterExitCallback(cbid);
  sceKernelSleepThreadCB();

  return 0;
}

// --------------------------------------------------------------------------

int SetupCallbacks(void) {
  int thid = 0;

  //thid = sceKernelCreateThread("update_thread", CallbackThread, 0x11, 0xFA0, PSP_THREAD_ATTR_USER, 0);
  thid = sceKernelCreateThread("update_thread", CallbackThread, 0x11, 0xFA0, 0, 0);
  if (thid >= 0) {
    sceKernelStartThread(thid, 0, 0);
  }

  return thid;
}

// --------------------------------------------------------------------------

void ColorPrintF(int color, char *fmt, ...) {
  va_list ap;
  char buffer[256];
  int size;

  va_start(ap, fmt);
  size = vsnprintf(buffer, 256, fmt, ap);
  pspDebugScreenSetTextColor(color);
  pspDebugScreenPrintData(buffer, size);
  pspDebugScreenSetTextColor(WHITE);
  va_end(ap);
}

// --------------------------------------------------------------------------

void CenterPrintF(int color, int offset, char *fmt, ...) {
  va_list ap;
  char buffer[256];
  int size;

  int dx = SCR_WIDTH;
  int dy = SCR_HEIGHT;

  va_start(ap, fmt);
  size = vsnprintf(buffer, 256, fmt, ap);
  pspDebugScreenSetXY((dx - strlen(buffer)) / 2, (dy / 2) - offset);
  pspDebugScreenSetTextColor(color);
  pspDebugScreenPrintData(buffer, size);
  pspDebugScreenSetTextColor(WHITE);
  va_end(ap);
}

// --------------------------------------------------------------------------

int ConnectToClient(int controller) {
  int sock;
  struct sockaddr_in addr;
  int err;
  char buf[MAXPATHLEN];
  char file_path[128];
  char Buffer[512];
  char *Scan;
  FILE *CfgFile;
  char host[32];
  unsigned short port;
  unsigned short port1 = PORT1;
  unsigned short port2 = PORT2;
  unsigned short port3 = PORT3;
  unsigned short port4 = PORT4;
  int found = 0;
  // char temp[10];

  switch (controller) {
    case 1:
      port = port1;
      break;
    case 2:
      port = port2;
      break;
    case 3:
      port = port3;
      break;
    case 4:
      port = port4;
      break;
    default:
      port = port1;
  }

  pspDebugScreenClear();

  strcpy(file_path, getcwd(buf, MAXPATHLEN));
  strcat(file_path, "/wifi.cfg");

  printf("Reading configuration file: %s \n", file_path);

  CfgFile = fopen(file_path, "r");

  if (CfgFile != (FILE*)0) {
    while (fgets(Buffer, 512, CfgFile) != (char *)0) {
      Scan = strchr(Buffer, '\n');
      if (Scan) *Scan = '\0';
      Scan = strchr(Buffer, '\r');
      if (Scan) *Scan = '\0';

      if (Buffer[0] == '#') continue;

      if (!strncasecmp(Buffer, "address=", 8)) {
        printf("Found IP address!\n");
        found = 1;
        strcpy(host, Buffer + 8);
        printf("Setting IP address to connect to as %s.\n", host);
      }
    }
    fclose(CfgFile);
    if (found == 0) {
      ColorPrintF(RED, "Could not find IP address in configuration file!");
      done = 1;
    }
  }

  printf("Controller selection is %d, we will be connecting to port %d.\n", controller, port);

  sceKernelDelayThread(3000 * 1000);

  memset(&addr, 0, sizeof(struct sockaddr_in));
  addr.sin_family = AF_INET;
  addr.sin_port = htons(port);
  sceNetInetInetAton(host, &addr.sin_addr);

  sock = sceNetInetSocket(AF_INET, SOCK_DGRAM, 0);
  if (sock < 0) {
    return -1;
  }

  err = sceNetInetConnect(sock, &addr, sizeof(addr));
  if (err) {
    return -1;
  }

  return sock;
}

// --------------------------------------------------------------------------

void read_pad(int controller) {
  int sock;
  int batteryLifeTime = 0;

  int dx = SCR_WIDTH;
  int dy = SCR_HEIGHT;

  int inverted = 0;

  SceCtrlData pad;

  unsigned int packetNum = 1;
  unsigned char packet[11];
  memset(packet, 0, sizeof(packet));

  sock = ConnectToClient(controller);
  if (sock < 0) {
    CenterPrintF(RED, 0, "Error opening socket!\n");
    return;
  }

  sceCtrlSetSamplingCycle(0);
  sceCtrlSetSamplingMode(PSP_CTRL_MODE_ANALOG);

  pspDebugScreenClear();

  while (!done) {
    if (sceWlanGetSwitchState() != 1) {
      connect_to_apctl(1);
      read_pad(controller);
    }
    sceCtrlReadBufferPositive(&pad, 1);

    if ((pad.Buttons & PSP_CTRL_LTRIGGER) && (pad.Buttons & PSP_CTRL_RTRIGGER) && (pad.Buttons & PSP_CTRL_SELECT)) {
      pspDebugScreenClear();
      if (inverted == 0) {
        pspDebugScreenSetXY( (dx - strlen(INVERTY)) / 2, dy / 2);
        ColorPrintF(RED, INVERTY);
        inverted = 1;
      } else if (inverted == 1) {
        pspDebugScreenSetXY( (dx - strlen(INVERTN)) / 2, dy / 2);
        ColorPrintF(RED, INVERTN);
        inverted = 0;
      }
      sceKernelDelayThread(2000 * 1000);
      pspDebugScreenClear();
    }

    pspDebugScreenSetXY((dx - (strlen(ANALOGX) + 3)) / 2, ((dy / 2) - 4));
    printf(ANALOGX "%d  \n", pad.Lx);
    pspDebugScreenSetXY((dx - (strlen(ANALOGY) + 3)) / 2, ((dy / 2) - 3));
    printf(ANALOGY "%d  \n", (inverted == 1 ? 255 - pad.Ly : pad.Ly));

    pspDebugScreenSetXY((dx - (strlen(LTRIGGER) + strlen(RTRIGGER))) / 2, ((dy / 2) - 2));
    ColorPrintF(pad.Buttons & PSP_CTRL_LTRIGGER ? BLUE : WHITE, LTRIGGER);
    ColorPrintF(pad.Buttons & PSP_CTRL_RTRIGGER ? BLUE : WHITE, RTRIGGER);

    pspDebugScreenSetXY((dx - (strlen(UP) + strlen(TRIANGLE))) / 2, ((dy / 2) - 1));
    ColorPrintF(pad.Buttons & PSP_CTRL_UP ? BLUE : WHITE, UP);
    ColorPrintF(pad.Buttons & PSP_CTRL_TRIANGLE ? BLUE : WHITE, TRIANGLE);

    pspDebugScreenSetXY((dx - (strlen(LEFT) + strlen(RIGHT) + strlen(SQUARE) + strlen(CIRCLE))) / 2, (dy / 2));
    ColorPrintF(pad.Buttons & PSP_CTRL_LEFT ? BLUE : WHITE, LEFT);
    ColorPrintF(pad.Buttons & PSP_CTRL_RIGHT ? BLUE : WHITE, RIGHT);
    ColorPrintF(pad.Buttons & PSP_CTRL_SQUARE ? BLUE : WHITE, SQUARE);
    ColorPrintF(pad.Buttons & PSP_CTRL_CIRCLE ? BLUE : WHITE, CIRCLE);

    pspDebugScreenSetXY((dx - (strlen(DOWN) + strlen(CROSS))) / 2, ((dy / 2) + 1));
    ColorPrintF(pad.Buttons & PSP_CTRL_DOWN ? BLUE : WHITE, DOWN);
    ColorPrintF(pad.Buttons & PSP_CTRL_CROSS ? BLUE : WHITE, CROSS);

    pspDebugScreenSetXY((dx - (strlen(SELECT) + strlen(START))) / 2, ((dy / 2) + 2));
    ColorPrintF(pad.Buttons & PSP_CTRL_SELECT ? BLUE : WHITE, SELECT);
    ColorPrintF(pad.Buttons & PSP_CTRL_START ? BLUE : WHITE, START);

    if (scePowerIsBatteryExist()) {
      batteryLifeTime = scePowerGetBatteryLifeTime();
      pspDebugScreenSetXY(0, (dy - 1));

      if (scePowerIsPowerOnline()) {
        printf("Running on external power...                                            ");
      } else {
        printf("Battery - Life: ");
        ColorPrintF(scePowerIsLowBattery() ? RED : WHITE, "%d", scePowerGetBatteryLifePercent());
        printf("%% (%02dh%02dm) / Temperature: %d C / Voltage: %0.3fV        ",
          batteryLifeTime / 60,
          batteryLifeTime - (batteryLifeTime / 60 * 60),
          scePowerGetBatteryTemp(),
          (float) scePowerGetBatteryVolt() / 1000.0);
      }
    }

    /* The code below was written by Mikko Monone. */
    unsigned char packet[11];
    packet[0] = '!';
    packet[1] = 'K';
    packet[2] = (packetNum) & 0xff;
    packet[3] = (packetNum >> 8) & 0xff;
    packet[4] = (packetNum >> 16) & 0xff;
    packet[5] = (packetNum >> 24) & 0xff;
    packet[6] = pad.Lx;
    packet[7] = (inverted == 1 ? 255 - pad.Ly : pad.Ly);
    packet[8] = (pad.Buttons) & 0xff;
    packet[9] = (pad.Buttons >> 8) & 0xff;
    packet[10] = (pad.Buttons >> 16) & 0xff;
    packetNum++;
    /* The code above was written by Mikko Monone. */
    sceNetInetSend(sock, packet, sizeof(packet), 0);
    sceDisplayWaitVblankStart();
  }
}

// --------------------------------------------------------------------------

int connect_to_apctl(int config) {
  int err;
  int stateLast = -1;
  char *stateinfo;

  if (sceWlanGetSwitchState() != 1)
    pspDebugScreenClear();

  while (sceWlanGetSwitchState() != 1) {
    pspDebugScreenSetXY(0, 0);
    ColorPrintF(RED, "Please enable WLAN to continue.\n");
    sceKernelDelayThread(1000 * 1000);
  }

  err = sceNetApctlConnect(config);
  if (err != 0) {
    CenterPrintF(RED, 0, MODULE_NAME ": sceNetApctlConnect returns %08X\n", err);
    return 0;
  }

  printf(MODULE_NAME ": Connecting to Access Point.\n");
  while (1) {
    int state;
    err = sceNetApctlGetState(&state);
    if (err != 0) {
      CenterPrintF(RED, 0, MODULE_NAME ": sceNetApctlGetState returns $%x\n", err);
      break;
    }
    if (state > stateLast) {
      switch (state) {
        case 0:
          stateinfo = "Idle";
          break;
        case 1:
          stateinfo = "Starting up";
          break;
        case 2:
          stateinfo = "Starting up";
          break;
        case 3:
          stateinfo = "Obtaining IP address";
          break;
        case 4:
          stateinfo = "IP address obtained";
          break;
        default:
          stateinfo = "Idle";
      }
      printf("  Connection state %d of 4: %s.\n", state, stateinfo);
      stateLast = state;
    }
    if (state == 4) {
      break;
    }

    sceKernelDelayThread(50 * 1000);
  }
  printf(MODULE_NAME ": Connected to Access Point!\n");
  sceKernelDelayThread(2000 * 1000);

  if (err != 0) {
    return 0;
  }

  return 1;
}

// --------------------------------------------------------------------------

char *getconfname(int confnum) {
  sceUtilityGetNetParam(confnum, PSP_NETPARAM_NAME, (netData *)confname);
  return confname;
}

// --------------------------------------------------------------------------

int net_thread(SceSize args, void *argp) {
  int err;
  int i;
  int start = 1;
  int selComponent = 1;
  int selController = 1;
  int select = 1;
  int last = 0;
  int lastC = 0;

  SceCtrlData pad;
  sceCtrlSetSamplingCycle(0);
  sceCtrlSetSamplingMode(PSP_CTRL_MODE_ANALOG);
  
  pspDebugScreenClear();

  do {
    if ((err = pspSdkInetInit())) {
      CenterPrintF(RED, 0, MODULE_NAME ": Error, could not initialize the network.  Error Code: %08X\n", err);
      break;
    }

    while (sceUtilityCheckNetParam(start) == 0) {
      start++;
    }
    start--;

    pspDebugScreenSetXY(0, 0);
    printf("%d internet configuration(s) found.\n\n", start);
    printf("Select which configuration to use from\n");
    printf("the list and press either %s or %s.\n", NSCROSS, NSCIRCLE);
    i = 1;
    while (i <= start) {
      printf("  Configuration %d: %s\n", i, getconfname(i));
      i++;
    }

    while (select) {
      sceCtrlReadBufferPositive(&pad, 1);
      if (pad.Buttons & PSP_CTRL_UP) {
        if (selComponent > 1) {
          selComponent--;
        }
        sceKernelDelayThread(100 * 1000);
      } else if (pad.Buttons & PSP_CTRL_DOWN) {
        if (selComponent < start) {
          selComponent++;
        }
        sceKernelDelayThread(100 * 1000);
      }
      if ((pad.Buttons & PSP_CTRL_CIRCLE) || (pad.Buttons & PSP_CTRL_CROSS)) {
        select = 0;
      }
      if (last != selComponent) {
        if (last == 0) { last = 1; }
        pspDebugScreenSetXY(0, 3 + last);
        ColorPrintF(WHITE, "  Configuration %d: %s", last, getconfname(last));

        pspDebugScreenSetXY(0, 3 + selComponent);
        ColorPrintF(RED, "  Configuration %d: %s", selComponent, getconfname(selComponent));
      }
      last = selComponent;
      sceDisplayWaitVblankStart();
    }

    pspDebugScreenSetXY(0, 4 + start);

    printf("\nUsing connection %d (%s) to connect...\n\n", selComponent, getconfname(selComponent));

    sceKernelDelayThread(1000 * 1000);

    pspDebugScreenClear();

    select = 1;

    pspDebugScreenSetXY(0, 0);
    printf("Select which controller you will be from\n");
    printf("the list and press either %s or %s.\n", NSCROSS, NSCIRCLE);
    i = 1;
    while (i <= 4) {
      printf("  Controller %d\n", i);
      i++;
    }

    while (select) {
      sceCtrlReadBufferPositive(&pad, 1);
      if (pad.Buttons & PSP_CTRL_UP) {
        if (selController > 1) {
          selController--;
        }
        sceKernelDelayThread(100 * 1000);
      } else if (pad.Buttons & PSP_CTRL_DOWN) {
        if (selController < 4) {
          selController++;
        }
        sceKernelDelayThread(100 * 1000);
      }
      if ((pad.Buttons & PSP_CTRL_CIRCLE) || (pad.Buttons & PSP_CTRL_CROSS)) {
        select = 0;
      }
      if (lastC != selController) {
        if (lastC == 0) { lastC = 1; }
        pspDebugScreenSetXY(0, 1 + lastC);
        ColorPrintF(WHITE, "  Controller %d", lastC);

        pspDebugScreenSetXY(0, 1 + selController);
        ColorPrintF(RED, "  Controller %d", selController);
      }
      lastC = selController;
      sceDisplayWaitVblankStart();
    }

    pspDebugScreenSetXY(0, 6);

    printf("\nActing as controller %d\n\n", selController);

    sceKernelDelayThread(1000 * 1000);

    pspDebugScreenClear();

    if (connect_to_apctl(selComponent)) {
      char szMyIPAddr[32];
      if (sceNetApctlGetInfo(8, szMyIPAddr) != 0)
        strcpy(szMyIPAddr, "unknown IP address");

      read_pad(selController);
    }
  }
  while (0);

  return 0;
}

// --------------------------------------------------------------------------

int main(void) {
  SceUID thid;

  SetupCallbacks();

  pspDebugScreenInit();

  CenterPrintF(WHITE, 1, "WiFiController v0.4.4");
  CenterPrintF(WHITE, 0, "Copyright (c)2007 David Rudie");
  sceKernelDelayThread(2000 * 1000);
  pspDebugScreenClear();

# ifdef PSPFW3
  sceUtilityLoadNetModule(1);
  sceUtilityLoadNetModule(3);
# else
  if (pspSdkLoadInetModules() < 0) {
    CenterPrintF(RED, 0, "Error!  inet modules could not be loaded!\n");
    sceKernelSleepThread();
  }
# endif

  thid = sceKernelCreateThread("net_thread", net_thread, 0x18, 0x10000, PSP_THREAD_ATTR_USER, NULL);
  if (thid < 0) {
    CenterPrintF(RED, 0, "Error! Thread could not be created!\n");
    sceKernelSleepThread();
  }

  sceKernelStartThread(thid, 0, NULL);

  sceKernelExitDeleteThread(0);

  return 0;
}


mikko_monone-code/pspcontroller.h

#ifndef __PSPCONTROLLER_H__
#define __PSPCONTROLLER_H__

enum PSPCtrlButtons
{
	PSP_CTRL_SELECT     = 0x000001,
	PSP_CTRL_START      = 0x000008,
	PSP_CTRL_UP         = 0x000010,
	PSP_CTRL_RIGHT      = 0x000020,
	PSP_CTRL_DOWN      	= 0x000040,
	PSP_CTRL_LEFT      	= 0x000080,
	PSP_CTRL_LTRIGGER   = 0x000100,
	PSP_CTRL_RTRIGGER   = 0x000200,
	PSP_CTRL_TRIANGLE   = 0x001000,
	PSP_CTRL_CIRCLE     = 0x002000,
	PSP_CTRL_CROSS      = 0x004000,
	PSP_CTRL_SQUARE     = 0x008000,
	PSP_CTRL_HOME       = 0x010000,
	PSP_CTRL_HOLD       = 0x020000,
	PSP_CTRL_NOTE       = 0x800000,
};

struct PSPCtrlData
{
	unsigned int		Buttons;
	unsigned char 	Lx;
	unsigned char 	Ly;
};

// Initialise the controller server
bool	PSPCtrlInit( int port );
void	PSPCtrlRead( PSPCtrlData& ctrl );
void	PSPCtrlEnd();


#endif

mikko_monone-code/main.cpp
//
// PSP Controller PC host
//		by Mikko Mononen 2005 <memon@inside.org>
//
// Description:
//		This program receives the status of the PSP pad sent at each screen update (60Hz).
//		The data is received as 16 byte packed via UDP. This program was designed to
//		let the PSP act as controller to create quick game prototypes on PC.
//
#include <windows.h>
#include <stdio.h>
#include <conio.h>
#include <winioctl.h>
#include "pspcontroller.h"
#include "ppjioctl.h"


#define	NUM_ANALOG	2							// Number of analog values which we will provide
#define	NUM_DIGITAL	15						// Number of digital values which we will provide


#pragma pack(push,1)							// All fields in structure must be byte aligned.
typedef struct
{
 unsigned long	signature;				// Signature to identify packet to PPJoy IOCTL
 char			numAnalog;							// Num of analog values we pass
 long			analog[NUM_ANALOG];			// Analog values
 char			numDigital;							// Num of digital values we pass
 char			digital[NUM_DIGITAL];		// Digital values
}	JOYSTICK_STATE;
#pragma pack(pop)


int main( int argc, char **argv )
{
	printf( "PSP Controller test PC host\nMikko Mononen 2005\n\nPress Esc to quit.\n\n" );

	HANDLE					h;
	char						ch;
	JOYSTICK_STATE	joyState;
	DWORD	retSize, rc;
	char*	devName;
	int		port;

	devName= "\\\\.\\PPJoyIOCTL1";
	if( argc >= 2 )
		devName = argv[1];

	port = 20001;
	if( argc >= 3 )
		port = atoi( argv[2] );

	// Open a handle to the control device for the first virtual joystick.
	// Virtual joystick devices are names PPJoyIOCTL1 to PPJoyIOCTL16.
	h = CreateFile( devName, GENERIC_WRITE, FILE_SHARE_WRITE, NULL, OPEN_EXISTING, 0, NULL );

	/* Make sure we could open the device! */
	if( h == INVALID_HANDLE_VALUE )
	{
		printf( "CreateFile failed with error code %d trying to open %s device\n", GetLastError(), devName );
		return 1;
	}

	// Initialise the IOCTL data structure
	joyState.signature = JOYSTICK_STATE_V1;
	joyState.numAnalog = NUM_ANALOG;
	joyState.numDigital = NUM_DIGITAL;

	if( !PSPCtrlInit( port ) )
	{
		printf( "Error initialising the PSPCtrl.\n" );
		return 1;
	}

	PSPCtrlData	ctrl;

	while( true )
	{
		if( _kbhit() )
		{
			if( getch() == 27 )
				break;
		}

		joyState.analog[0] = joyState.analog[1] = (PPJOY_AXIS_MIN + PPJOY_AXIS_MAX) / 2;
		memset( joyState.digital, 0, sizeof( joyState.digital ) );

		PSPCtrlRead( ctrl );

		int	x = (int)ctrl.Lx - 128;
		int	y = (int)ctrl.Ly - 128;

		// There is a deadzone at the center of the controller. The readins at that area are not reliable.
		const int PSP_DEAD_ZONE = 14;
		const int	PSP_RANGE = 127 - PSP_DEAD_ZONE;

		if( x <= -PSP_DEAD_ZONE )			x += PSP_DEAD_ZONE;
		else if( x >= PSP_DEAD_ZONE )	x -= PSP_DEAD_ZONE;
		else x = 0;
		if( x < -PSP_RANGE ) x = -PSP_RANGE;

		if( y <= -PSP_DEAD_ZONE )			y += PSP_DEAD_ZONE;
		else if( y >= PSP_DEAD_ZONE )	y -= PSP_DEAD_ZONE;
		else y = 0;
		if( y < -PSP_RANGE ) y = -PSP_RANGE;

		joyState.analog[0] = PPJOY_AXIS_MIN + (x + PSP_RANGE) * (PPJOY_AXIS_MAX - PPJOY_AXIS_MIN) / (2 * PSP_RANGE);
		joyState.analog[1] = PPJOY_AXIS_MIN + (y + PSP_RANGE) * (PPJOY_AXIS_MAX - PPJOY_AXIS_MIN) / (2 * PSP_RANGE);

		joyState.digital[0] = (ctrl.Buttons & PSP_CTRL_TRIANGLE) ? 1 : 0;
		joyState.digital[1] = (ctrl.Buttons & PSP_CTRL_CIRCLE) ? 1 : 0;
		joyState.digital[2] = (ctrl.Buttons & PSP_CTRL_CROSS) ? 1 : 0;
		joyState.digital[3] = (ctrl.Buttons & PSP_CTRL_SQUARE) ? 1 : 0;

		joyState.digital[4] = (ctrl.Buttons & PSP_CTRL_LTRIGGER) ? 1 : 0;
		joyState.digital[5] = (ctrl.Buttons & PSP_CTRL_RTRIGGER) ? 1 : 0;

		joyState.digital[6] = (ctrl.Buttons & PSP_CTRL_UP) ? 1 : 0;
		joyState.digital[7] = (ctrl.Buttons & PSP_CTRL_RIGHT) ? 1 : 0;
		joyState.digital[8] = (ctrl.Buttons & PSP_CTRL_DOWN) ? 1 : 0;
		joyState.digital[9] = (ctrl.Buttons & PSP_CTRL_LEFT) ? 1 : 0;

		joyState.digital[10] = (ctrl.Buttons & PSP_CTRL_SELECT) ? 1 : 0;
		joyState.digital[11] = (ctrl.Buttons & PSP_CTRL_START) ? 1 : 0;
		joyState.digital[12] = (ctrl.Buttons & PSP_CTRL_HOME) ? 1 : 0;
		joyState.digital[13] = (ctrl.Buttons & PSP_CTRL_HOLD) ? 1 : 0;
		joyState.digital[14] = (ctrl.Buttons & PSP_CTRL_NOTE) ? 1 : 0;

		// Send request to PPJoy for processing.
		// Currently there is no Return Code from PPJoy, this may be added at a
		// later stage. So we pass a 0 byte output buffer.
		if( !DeviceIoControl( h, IOCTL_PPORTJOY_SET_STATE, &joyState, sizeof( joyState ), NULL, 0, &retSize, NULL ) )
		{
			rc = GetLastError();
			if( rc == 2 )
			{
				printf( "Underlying joystick device deleted. Exiting read loop\n" );
				break;
			}
			printf( "DeviceIoControl error %d\n",rc );
		}

		Sleep( 10 );
	}	

	PSPCtrlEnd();
	CloseHandle( h );

	return 0;
}